test/Mono.Linker.Tests.Cases/DataFlow/NullableAnnotations.cs

```diff
@@ -0,0 +1,289 @@
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2087")]
+		static void UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TNullable> ()
+		{
+			(Nullable.GetUnderlyingType (typeof (TNullable))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void NullableOfAnnotatedGenericParameterRequiresPublicProperties<[KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] T> () where T : struct
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<T>))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		[ExpectedWarning ("IL2087")]
+		static void NullableOfUnannotatedGenericParameterRequiresPublicProperties<T> () where T : struct
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<T>))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void NullableOfAnnotatedParameterRequiresPublicProperties ([KeptAttributeAttribute (typeof (DAM))][DAM (DAMT.PublicProperties)] Type t)
+		{
+			(Nullable.GetUnderlyingType (typeof (Nullable<>).MakeGenericType (t))).RequiresPublicProperties ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnClasses ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (TestClass));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<TestClass> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (TestClass));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TestClass> ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnStructs ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (TestStruct));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<TestStruct> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (TestStruct));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<TestStruct> ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOnNullableStructs ()
+		{
+			UnderlyingTypeOfAnnotatedParameterRequiresPublicProperties (typeof (Nullable<TestStruct>));
+			UnderlyingTypeOfAnnotatedGenericParameterRequiresPublicProperties<Nullable<TestStruct>> ();
+			UnderlyingTypeOfUnannotatedParameterRequiresPublicProperties (typeof (Nullable<TestStruct>));
+			UnderlyingTypeOfUnannotatedGenericParameterRequiresProperties<Nullable<TestStruct>> ();
+		}
+
+		[Kept]
+		static void TestGetUnderlyingTypeOfCreatedNullableOnStructs ()
+		{
+			NullableOfAnnotatedParameterRequiresPublicProperties (typeof (TestStruct));
+			NullableOfUnannotatedGenericParameterRequiresPublicProperties<TestStruct> ();
+			NullableOfUnannotatedParameterRequiresPublicProperties (typeof (TestStruct));
+			NullableOfUnannotatedGenericParameterRequiresPublicProperties<TestStruct> ();
+		}
+
+		[Kept]
+		// Bug - GetUnderlyingType intrinsic handling does not yet have special handling of Nullables
+		//[ExpectedWarning("IL2067")]
+		static void NullableOfUnannotatedParameterRequiresPublicProperties (Type t)
```

Comment:
Exactly the case I used as my counter example above... I think we need to fix this.